### ISSUE TITLE
Integration of DdiMon into GuardMon

### ISSUE THREAD

**Issue by ghost** [tandasat/GuardMon#5]
Hello @tandasat!
Firstly thank you for all your worth knowing informations about the HV topic.
I wanted you to ask if its possible to integrate DdiMon into GuardMon?
I tried it already, and I think I implemented it the right way, but sadly I always get a BSOD after:
```
21:19:57.449	INF	#5	    4	 6344	System         	Log has been initialized.
21:19:57.457	INF	#0	    4	 6344	System         	Initializing VMX for the processor 0.
21:19:57.872	INF	#0	    4	 6344	System         	Initialized successfully.
21:19:57.874	INF	#1	    4	 6344	System         	Initializing VMX for the processor 1.
21:19:58.290	INF	#1	    4	 6344	System         	Initialized successfully.
21:19:58.292	INF	#2	    4	 6344	System         	Initializing VMX for the processor 2.
21:19:58.713	INF	#2	    4	 6344	System         	Initialized successfully.
21:19:58.715	INF	#3	    4	 6344	System         	Initializing VMX for the processor 3.
21:19:59.130	INF	#3	    4	 6344	System         	Initialized successfully.
21:19:59.131	INF	#4	    4	 6344	System         	Initializing VMX for the processor 4.
21:19:59.545	INF	#4	    4	 6344	System         	Initialized successfully.
21:19:59.547	INF	#5	    4	 6344	System         	Initializing VMX for the processor 5.
21:19:59.961	INF	#5	    4	 6344	System         	Initialized successfully.
21:19:59.962	INF	#6	    4	 6344	System         	Initializing VMX for the processor 6.
21:20:00.376	INF	#6	    4	 6344	System         	Initialized successfully.
21:20:00.376	INF	#7	    4	 6344	System         	Initializing VMX for the processor 7.
21:20:00.792	INF	#7	    4	 6344	System         	Initialized successfully.
21:20:00.793	INF	#7	    4	 6344	System         	Hook has been installed at fffff8029e582010 ExAllocatePoolWithTag.
21:20:00.796	INF	#6	    4	 6344	System         	Hook has been installed at fffff8029e5839e0 ExFreePool.
21:20:00.797	INF	#5	    4	 6344	System         	Hook has been installed at fffff8029e582c60 ExFreePoolWithTag.
21:20:00.800	INF	#5	    4	 6344	System         	Hook has been installed at fffff8029e36f530 ExQueueWorkItem.
21:20:00.802	INF	#5	    4	 6344	System         	Hook has been installed at fffff8029e76f4b0 NtQuerySystemInformation.
```
BSOD Code: KMODE_EXCEPTION_NOT_HANDLED

If I want to do that, which HyperPlatform version I should take? 
The one from DdiMon or from GuardMon?
I ported your code exactly as you created it into the right areas:
```
DdimonInitialization()
  DdimonpEnumExportedSymbolsCallback()  // Enumerates exports of ntoskrnl
    ShInstallHook()                     // Installs a stealth hook
  ShEnableHooks()                       // Activates installed hooks
    ShEnablePageShadowing()
      ShpEnablePageShadowingForExec()   // Configures an EPT entry as
                                        // explained in "Default state"
```
into the GuardMon code, it compiles, but sadly always a BSOD.
If I shall post the MEMORY.DMP let me know please, but I didn't really saw something bad in there.
Thanks for your assistance!

**Comment by tandasat** [comment]
I would start with the version in DdiMon since changes for GuardMon is rather simple and easy to port (it is a [single commit](https://github.com/tandasat/HyperPlatform/compare/guardmon) so you might be able to cherry-pick it easily).

Also, since I do not very actively maintain those project for the latest version of Windows, I would make sure the latest version of DdiMon works as expected on your test system before using it as a base. (If does not work, please file a ticket.)

As for the bug check, I can try to guess how it went wrong if you could upload the memory dump, and a pdb file of the build if you are OK with this.

**Comment by ghost** [comment]
> 
> 
> I would start with the version in DdiMon since changes for GuardMon is rather simple and easy to port (it is a [single commit](https://github.com/tandasat/HyperPlatform/compare/guardmon) so you might be able to cherry-pick it easily).
> 
> Also, since I do not very actively maintain those project for the latest version of Windows, I would make sure the latest version of DdiMon works as expected on your test system before using it as a base. (If does not work, please file a ticket.)
> 
> As for the bug check, I can try to guess how it went wrong if you could upload the memory dump, and a pdb file of the build if you are OK with this.

I've tested it before using it as a base, and it works as intended on 1803.
Sure thing, here is the MEMORY.DMP & the .PDB of the latest build:

**Comment by tandasat** [comment]
It does not appear that you are intercepting int3 (#BP). Make sure you have properly ported DdiMon and try again.

The #BP is triggered at `ExAllocatePoolWithTag`. The address does not contain int3 so you have properly installed stealth breakpoint. However, the exception is not handled by the hypervisor as it should be, hence, the exception is delivered to and unhandled by the OS (since there is no SEH for the API), leading to the bug check.
```
5: kd> k
 # Child-SP          RetAddr           Call Site
00 fffffa8c`5e745558 fffff802`9e4d9903 nt!KeBugCheckEx
...
07 fffffa8c`5e746420 fffff802`9e44bfd2 nt!KiExceptionDispatch+0xc2
08 fffffa8c`5e746600 fffff802`9e582011 nt!KiBreakpointTrap+0x2d2
09 fffffa8c`5e746798 fffff802`9e828348 nt!ExAllocatePoolWithTag+0x1

5: kd> u nt!ExAllocatePoolWithTag+0x1
nt!ExAllocatePoolWithTag+0x1:
fffff802`9e582011 push    rbp
fffff802`9e582012 push    r12
fffff802`9e582014 push    r13
fffff802`9e582016 push    r14
```

You can find relevant changes you might be missing here:
https://github.com/tandasat/HyperPlatform/compare/ddimon#diff-8823c34c41b7fbb77a5212b8f130e919R637
https://github.com/tandasat/HyperPlatform/compare/ddimon#diff-8aa539026731c1fd7eebbb17cc8ae848R392

**Comment by ghost** [comment]
> 
> 
> It does not appear that you are intercepting int3 (#BP). Make sure you have properly ported DdiMon and try again.
> 
> The #BP is triggered at `ExAllocatePoolWithTag`. The address does not contain int3 so you have properly installed stealth breakpoint. However, the exception is not handled by the hypervisor as it should be, hence, the exception is delivered to and unhandled by the OS (since there is no SEH for the API), leading to the bug check.
> 
> ```
> 5: kd> k
>  # Child-SP          RetAddr           Call Site
> 00 fffffa8c`5e745558 fffff802`9e4d9903 nt!KeBugCheckEx
> ...
> 07 fffffa8c`5e746420 fffff802`9e44bfd2 nt!KiExceptionDispatch+0xc2
> 08 fffffa8c`5e746600 fffff802`9e582011 nt!KiBreakpointTrap+0x2d2
> 09 fffffa8c`5e746798 fffff802`9e828348 nt!ExAllocatePoolWithTag+0x1
> 
> 5: kd> u nt!ExAllocatePoolWithTag+0x1
> nt!ExAllocatePoolWithTag+0x1:
> fffff802`9e582011 push    rbp
> fffff802`9e582012 push    r12
> fffff802`9e582014 push    r13
> fffff802`9e582016 push    r14
> ```
> 
> You can find relevant changes you might be missing here:
> https://github.com/tandasat/HyperPlatform/compare/ddimon#diff-8823c34c41b7fbb77a5212b8f130e919R637
> https://github.com/tandasat/HyperPlatform/compare/ddimon#diff-8aa539026731c1fd7eebbb17cc8ae848R392

It seems the Issue was:
```
const auto exception_bitmap =
//1 << InterruptionVector::kBreakpointException |
1 << InterruptionVector::kBreakpointException |
// 1 << InterruptionVector::kGeneralProtectionException |
// 1 << InterruptionVector::kPageFaultException |
0;
```

Now im using:
```
const auto exception_bitmap =
// 1 << InterruptionVector::kBreakpointException |
// 1 << InterruptionVector::kGeneralProtectionException |
1 << InterruptionVector::kGeneralProtectionException |
// 1 << InterruptionVector::kPageFaultException |
0;
```

Im now using the exception_bitmap "InterruptionVector::kGeneralProtectionException" instead of "InterruptionVector::kBreakpointException" which i've used before.
Which is working just fine it seems.
Thanks for your help!

**Comment by tandasat** [comment]
By the way, you can use multi-line code (quotes) with ```

It is hard to parse out some of your quotes :)

**Comment by ghost** [comment]
> 
> 
> By the way, you can use multi-line code (quotes) with ```
> 
> It is hard to parse out some of your quotes :)

Thank you, I didn't know that. I've edited my posts.
Have a nice day!
